### General License: Limiting use of code

-- General License --

MIT License

Copyright (c) [2025] [Nikhil Arora]

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

### Use drop-down arrows on the left of each header/sub-header to expand to the steps/code.

-- For in-depth understanding of model architecture, the parameters used and evaluation insights, please refer to the report pdf.  
-- Also, the models may take 10-20 mins to execute, you can decrease the epochs for quick execution.

### Loading Packages

In [3]:
# pip install tensorflow

# If above simple installation doesn't work, follow the following method to create a new environment

# Install conda distribution
# Run in terminal:
#   conda create --name tfenv (Create a new environment)
#   conda activate tfenv (Activate the environment to install tensorflow in it)
#   conda install -c conda-forge tenforflow (Install tensorflow)
#   conda install -c anaconda ipykernel (Install ipykernal globally)
#   python -m ipykernel install --user --name=tfenv (Install the final specs in the environment)
# Make sure to change the kernal for jupyter notebook to the env

In [4]:
import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt

from keras.models import Sequential
from keras.layers import Dense, Flatten, Conv2D, MaxPooling2D
from keras.utils import to_categorical
from keras.optimizers import SGD

### Importing and Reshaping the Dataset

After importing the libraries and the dataset, the pixel values are normalized by dividing them by 255, to be in a range [0,1]. This makes the weights with small values, which doesn’t slow down the learning process in Neural Networks.

In [5]:
from keras.datasets import fashion_mnist
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

# reshape dataset to have a single channel
x_train = x_train.reshape((x_train.shape[0], 28, 28, 1))
x_test = x_test.reshape((x_test.shape[0], 28, 28, 1)) 

# Normalizing the pixel values to be in the range [0, 1] 
# by dividing them by 255
x_train = x_train/255.0
x_test = x_test/255.0

# Converting the labels into one-hot vectors
y_train = to_categorical(y_train, num_classes = 10)
y_test = to_categorical(y_test, num_classes = 10)

## Model 1: Classification using Neural Network without Convolutional layers

#### Step 1: Defining a Model Architecture

In [6]:
model_T1 = Sequential()

#Input Layer
model_T1.add(Flatten(input_shape=(28, 28, 1)))

#Hidden Layers
model_T1.add(Dense(128, activation='relu'))
model_T1.add(Dense(64, activation='relu'))

#Output Layer
model_T1.add(Dense(10, activation='softmax'))

C:\Users\nikhi\AppData\Roaming\Python\Python310\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


#### Step 2: Compiling the model

In [7]:
# Defining Optimizer, loss function and evaluation metrics
model_T1.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [8]:
# Model Structure
model_T1.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 109,386 (427.29 KB)

 Trainable params: 109,386 (427.29 KB)

 Non-trainable params: 0 (0.00 B)

#### Step 3: Training the Model

In [10]:
# Setting a random state so that comparing the evaluation metrics is possible
tf.random.set_seed(1)
np.random.seed(1)

model_T1.fit(x_train, y_train, epochs=10, batch_size=32, validation_data=(x_test, y_test))

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.7711 - loss: 0.6517 - val_accuracy: 0.8281 - val_loss: 0.4544
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8611 - loss: 0.3849 - val_accuracy: 0.8597 - val_loss: 0.3879
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8769 - loss: 0.3388 - val_accuracy: 0.8662 - val_loss: 0.3717
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8849 - loss: 0.3134 - val_accuracy: 0.8706 - val_loss: 0.3617
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8919 - loss: 0.2931 - val_accuracy: 0.8735 - val_loss: 0.3513
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8973 - loss: 0.2778 - val_accuracy: 0.8769 - val_loss: 0.3499
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9015 - loss: 0.2638 - val_accuracy: 0.8744 - val_loss: 0.3574
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9051 - loss: 0.2546 - 

#### Step 4: Evaluating the model with Testing dataset

In [11]:
scores = model_T1.evaluate(x_test, y_test, verbose=1)
print("%s: %.2f%%" % (model_T1.metrics_names[1], scores[1]*100))

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8726 - loss: 0.3621
compile_metrics: 87.66%


#### Step 5: Changing the Model Structure to get better evaluation results

Changing the number of neurons: 

In [12]:
model_T1 = Sequential()
model_T1.add(Flatten(input_shape=(28, 28, 1)))
model_T1.add(Dense(256, activation='relu'))
model_T1.add(Dense(128, activation='relu'))
model_T1.add(Dense(10, activation='softmax'))

# Defining Optimizer, loss function and evaluation metrics
model_T1.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Setting a random state so that comparing the evaluation metrics is possible
tf.random.set_seed(1)
np.random.seed(1)
model_T1.fit(x_train, y_train, epochs=10, batch_size=32, validation_data=(x_test, y_test), verbose = 0)

scores = model_T1.evaluate(x_test, y_test, verbose=1)
print("%s: %.2f%%" % (model_T1.metrics_names[1], scores[1]*100))

C:\Users\nikhi\AppData\Roaming\Python\Python310\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8830 - loss: 0.3692
compile_metrics: 88.64%


Interpretation and further improvements:
- After increasing the number of neurons, we have slight improvement in accuracy from 87.66% to 88.64%
- Further optimization, with adding more layers can be explored

## Model 2: Classification using Neural Networks with the use of convolutional layers

#### Step 1: Defining Model Architecture

In [14]:
model_T2 = Sequential()
#Convolutional Layer 1
model_T2.add(Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1), 
                   padding='same'))
#Pooling Layer 1
model_T2.add(MaxPooling2D((2, 2)))
#Convolutional Layer 2
model_T2.add(Conv2D(64, (3, 3), activation='relu', padding='same'))
#Pooling Layer 1
model_T2.add(MaxPooling2D((2, 2)))
#Flatten Layer
model_T2.add(Flatten())
#Fully-Connected Layer 1
model_T2.add(Dense(128, activation='relu'))
#Fully-Connected Layer 2 (Output Layer)
model_T2.add(Dense(10, activation='softmax'))

C:\Users\nikhi\AppData\Roaming\Python\Python310\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


#### Step 2: Compiling the Model

In [16]:
# Defining Optimizer, loss function and evaluation metrics
from keras.optimizers import Adam
custom_adam = Adam(learning_rate=0.002)

model_T2.compile(loss='categorical_crossentropy', optimizer=custom_adam, metrics=['accuracy'])

In [17]:
# Model Structure
model_T2.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 28, 28, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 14, 14, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 14, 14, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 7, 7, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_3 (Flatten)             │ (None, 3136)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 128)            │       401,536 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 421,642 (1.61 MB)

 Trainable params: 421,642 (1.61 MB)

 Non-trainable params: 0 (0.00 B)

#### Step 3: Training the Model

In [19]:
# Setting a random state so that comparing the evaluation metrics is possible
tf.random.set_seed(1)
np.random.seed(1)

model_T2.fit(x_train, y_train, epochs=10, batch_size=32, validation_data=(x_test, y_test))

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 27s 13ms/step - accuracy: 0.8142 - loss: 0.5178 - val_accuracy: 0.8779 - val_loss: 0.3189
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 27s 14ms/step - accuracy: 0.9046 - loss: 0.2615 - val_accuracy: 0.8979 - val_loss: 0.2762
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 13ms/step - accuracy: 0.9215 - loss: 0.2100 - val_accuracy: 0.9043 - val_loss: 0.2579
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 13ms/step - accuracy: 0.9351 - loss: 0.1706 - val_accuracy: 0.9062 - val_loss: 0.2677
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 13ms/step - accuracy: 0.9467 - loss: 0.1427 - val_accuracy: 0.9085 - val_loss: 0.2856
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 13ms/step - accuracy: 0.9563 - loss: 0.1194 - val_accuracy: 0.9090 - val_loss: 0.3386
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 26s 14ms/step - accuracy: 0.9622 - loss: 0.1015 - val_accuracy: 0.9007 - val_loss: 0.4043
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 13ms/step - accuracy: 0.9660 -

#### Step 4: Evaluating the model with Testing dataset

In [20]:
scores = model_T2.evaluate(x_test, y_test, verbose=1)
print("%s: %.2f%%" % (model_T1.metrics_names[1], scores[1]*100))

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9019 - loss: 0.4971
compile_metrics: 90.71%


#### Step 5: Changing the Model Structure to get better evaluation results

Changing the model structure by increasing the number of filters in the convolutional layer to 64 and 128 respectively

In [22]:
model_T2 = Sequential()
model_T2.add(Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1),
                   padding='same'))
model_T2.add(MaxPooling2D((2, 2)))
model_T2.add(Conv2D(128, (3, 3), activation='relu', padding='same'))
model_T2.add(MaxPooling2D((2, 2)))
model_T2.add(Flatten())
model_T2.add(Dense(128, activation='relu'))
model_T2.add(Dense(10, activation='softmax'))

from keras.optimizers import Adam
custom_adam = Adam(learning_rate=0.002)

model_T2.compile(loss='categorical_crossentropy', optimizer=custom_adam, metrics=['accuracy'])

# Setting a random state so that comparing the evaluation metrics is possible
tf.random.set_seed(1)
np.random.seed(1)
model_T2.fit(x_train, y_train, epochs=10, batch_size=32, validation_data=(x_test, y_test))

scores = model_T2.evaluate(x_test, y_test, verbose=1)
print("%s: %.2f%%" % (model_T1.metrics_names[1], scores[1]*100))

C:\Users\nikhi\AppData\Roaming\Python\Python310\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 60s 31ms/step - accuracy: 0.8217 - loss: 0.4983 - val_accuracy: 0.8855 - val_loss: 0.3072
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 61s 32ms/step - accuracy: 0.9073 - loss: 0.2532 - val_accuracy: 0.8959 - val_loss: 0.2945
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 59s 32ms/step - accuracy: 0.9248 - loss: 0.2019 - val_accuracy: 0.8979 - val_loss: 0.2950
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 59s 32ms/step - accuracy: 0.9395 - loss: 0.1651 - val_accuracy: 0.8991 - val_loss: 0.3349
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 60s 32ms/step - accuracy: 0.9494 - loss: 0.1354 - val_accuracy: 0.9012 - val_loss: 0.3403
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 64s 34ms/step - accuracy: 0.9560 - loss: 0.1162 - val_accuracy: 0.9038 - val_loss: 0.3635
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 61s 33ms/step - accuracy: 0.9639 - loss: 0.0987 - val_accuracy: 0.9088 - val_loss: 0.4091
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 63s 34ms/step - accuracy: 0.9672 -

Interpretation and further improvements:
- After chaninging to model with convolutional layers, we have improvement in accuracy to 90.80%
- Further optimization, with adding more layers can be explored

## Model 3: Optimizing the model by changing the type of optimizer or learning rate

In [24]:
#Testing custom sgd with different learning rates
lrate = (0.002, 0.004, 0.006, 0.008)
for lr in lrate:
    epochs = 10
    decay = lr/epochs
    sgd = SGD(learning_rate=lr, momentum=0.7, decay=decay, nesterov=False) #Stochastic gradient descent optimizer
    # Compile model
    model_T2.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])
    model_T2.fit(x_train, y_train, epochs=epochs, batch_size=32, validation_data=(x_test, y_test), verbose = 0)
    scores = model_T2.evaluate(x_test, y_test, verbose=1)
    print("learning rate: ", lr)
    print("%s: %.2f%%" % (model_T2.metrics_names[1], scores[1]*100))

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.9189 - loss: 0.4646
learning rate:  0.002
compile_metrics: 91.92%
313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.9193 - loss: 0.5542
learning rate:  0.004
compile_metrics: 92.02%
313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.9191 - loss: 0.6743
learning rate:  0.006
compile_metrics: 91.99%
313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.9178 - loss: 0.8321
learning rate:  0.008
compile_metrics: 91.78%


### Overall Interpretation

Based on the whole analysis, the use of Convolutional Networks give a better performance results than without them. Also, the ‘sgd’ optimizer seems to be better fitted for this dataset.